agent/cache/watch.go

```diff
@@ -106,148 +106,192 @@ func (c *Cache) notifyBlockingQuery(ctx context.Context, r getOptions, correlati
-			// we can definitely avoid it.
-			wait += lib.RandomStagger(r.Info.MaxAge / 16)
-		}
+	// Reset or increment failure counter
+	if err == nil {
+		failures = 0
+	} else {
+		failures++
+		c.options.Logger.
+			With("error", err).
+			With("cache-type", r.TypeEntry.Name).
+			With("index", index).
+			Warn("handling error in Cache.Notify")
+	}
 
-		select {
-		case <-time.After(wait):
-		case <-ctx.Done():
-			return
+	var wait time.Duration
+	// Determining how long to wait before the next poll is complicated.
+	// First off the happy path and the error path waits are handled distinctly
+	//
+	// Once fetching the data through the cache returns an error (and until a
+	// non-error value is returned) the wait time between each round of the loop
+	// gets controlled by the backOffWait function. Because we would have waited
+	// at least until the age of the cached data was too old the error path should
+	// immediately retry the fetch and backoff on the time as needed for persistent
+	// failures which potentially will wait much longer than the MaxAge of the request
+	//
+	// When on the happy path we just need to fetch from the cache often enough to ensure
+	// that the data is not older than the MaxAge. Therefore after fetching the data from
+	// the cache we can sleep until the age of that data would exceed the MaxAge. Sometimes
+	// this will be for the MaxAge duration (like when only a single notify was executed so
+	// only 1 go routine is keeping the cache updated). Other times this will be some smaller
+	// duration than MaxAge (when multiple notify calls were executed and this go routine just
+	// got data back from the cache that was a cache hit after the other go routine fetched it
+	// without a hit). We cannot just set MustRevalidate on the request and always sleep for MaxAge
+	// as this would eliminate the single-flighting of these requests in the cache and
+	// the efficiencies gained by it.
+	if failures > 0 {
+		wait = backOffWait(failures)
+	} else {
+		// Calculate when the cached data's Age will get too stale and
+		// need to be re-queried. When the data's Age already exceeds the
+		// maxAge the pollWait value is left at 0 to immediately re-poll
+		if meta.Age <= r.Info.MaxAge {
+			wait = r.Info.MaxAge - meta.Age
 		}
+
+		// Add a small amount of random jitter to the polling time. One
+		// purpose of the jitter is to ensure that the next time
+		// we fetch from the cache the data will be stale (unless another
+		// notify go routine has updated it while this one is sleeping).
+		// Without this it would be possible to wake up, fetch the data
+		// again where the age of the data is strictly equal to the MaxAge
+		// and then immediately have to re-fetch again. That wouldn't
+		// be terrible but it would expend a bunch more cpu cycles when
+		// we can definitely avoid it.
+		wait += lib.RandomStagger(r.Info.MaxAge / 16)
 	}
+
+	select {
+	case <-time.After(wait):
+	case <-ctx.Done():
```

Comment:
What happen if we call a cancel on a done context does it add an error to the context or have no effect?

Reply:
From the stdlib docs:
> A CancelFunc tells an operation to abandon its work. A CancelFunc does not wait for the work to stop. A CancelFunc may be called by multiple goroutines simultaneously. After the first call, subsequent calls to a CancelFunc do nothing.